src/main/java/com/conveyal/datatools/manager/models/transform/PreserveCustomFieldsTransformation.java

```diff
@@ -0,0 +1,129 @@
+            lookup.put(hashKey, nextLine);
+        }
+        return lookup;
+    }
+
+    @Override
+    public void transform(FeedTransformZipTarget zipTarget, MonitorableJob.Status status) {
+        String tableName = table + ".txt";
+        Path targetZipPath = Paths.get(zipTarget.gtfsFile.getAbsolutePath());
+        Table specTable = Arrays.stream(Table.tablesInOrder)
+                .filter(t -> t.name.equals(table))
+                .findFirst()
+                .get();
+
+        try (FileSystem targetZipFs = FileSystems.newFileSystem(targetZipPath, (ClassLoader) null)) {
+            if (specTable == null) throw new Exception(String.format("could not find specTable for table %s", table));
+
+            List<String> specTableFields = specTable.specFields().stream().map(f -> f.name).collect(Collectors.toList());
+            List<String> tablePrimaryKeys = specTable.getPrimaryKeyNames();
+
+            Path targetTxtFilePath = getTablePathInZip(tableName, targetZipFs);
+
+            final File tempFile = File.createTempFile(tableName + "-temp", ".txt");
+            File output = File.createTempFile(tableName + "-output-temp", ".txt");
+            int rowsModified = 0;
+            List<String> customFields;
+
+            try (
+                InputStream is = Files.newInputStream(targetTxtFilePath);
+                CsvMapReader customFileReader = new CsvMapReader(new StringReader(csvData), CsvPreference.STANDARD_PREFERENCE);
+                CsvMapReader editorFileReader = new CsvMapReader(new InputStreamReader(is), CsvPreference.STANDARD_PREFERENCE);
+                CsvMapWriter writer = new CsvMapWriter(new FileWriter(output), CsvPreference.STANDARD_PREFERENCE);
+            ){
+
+                String[] customHeaders = customFileReader.getHeader(true);
+                String[] editorHeaders = editorFileReader.getHeader(true);
+
+                customFields = Arrays.stream(customHeaders).filter(h -> !specTableFields.contains(h)).collect(Collectors.toList());
+                if (customFields.isEmpty()) return;
+                String[] fullHeaders = ArrayUtils.addAll(editorHeaders, customFields.toArray(new String[0]));
+
+                HashMap<String, Map<String, String>> customFieldsLookup = createCsvHashMap(customFileReader, customHeaders, tablePrimaryKeys);
+                writer.writeHeader(fullHeaders);
+
+                Map<String, String> row;
+                while ((row = editorFileReader.read(editorHeaders)) != null) {
+                    List<String> editorCsvPrimaryKeyValues = tablePrimaryKeys.stream()
+                        .map(row::get)
+                        .collect(Collectors.toList());
+
+                    String hashKey = StringUtils.join(editorCsvPrimaryKeyValues, "_");
+                    Map<String, String> customCsvValues = customFieldsLookup.get(hashKey);
+                    Map<String, String> finalRow = row;
+                    customFields.stream().forEach(customField -> {
```

Comment:
(SonarLint suggestion)
```suggestion
                    customFields.forEach(customField -> {
```

Reply:
Updated in https://github.com/ibi-group/datatools-server/pull/536/commits/73ffc88304dd672662ef0922a67ed5aace2b1e62